File: packages/pvm-assemblyscript/assembly/parser.ts
import { decodeBlob } from './codec'
import { PVMInstruction } from './pvm'

/**
 * Parse result structure
 */
export class ParseResult {
  success: bool
  instructions: PVMInstruction[]
  bitmask: Uint8Array
  jumpTable: u32[]
  errors: string[]
  codeLength: u32

  constructor(
    success: bool,
    instructions: PVMInstruction[],
    bitmask: Uint8Array,
    jumpTable: u32[],
    errors: string[],
    codeLength: u32,
  ) {
    this.success = success
    this.instructions = instructions
    this.bitmask = bitmask
    this.jumpTable = jumpTable
    this.errors = errors
    this.codeLength = codeLength
  }
}

/**
 * PVM Parser implementation
 * Parses PVM instruction data and program blobs according to Gray Paper specification
 *
 * Gray Paper Reference: pvm.tex sections 7.1-7.3
 */
export class PVMParser {
  /**
   * Skip function Fskip(i) - determines distance to next instruction
   *
   * Gray Paper Equation 7.1:
   * Fskip(i) = min(24, j ∈ N : (k ∥ {1,1,.})_{i+1+j} = 1)
   *
   * @param instructionIndex - Index of instruction opcode in instruction data
   * @param opcodeBitmask - Bitmask indicating valid instruction boundaries
   * @returns Number of octets minus 1 to next instruction's opcode
   */
  skip(instructionIndex: i32, opcodeBitmask: Uint8Array): i32 {
    // Append bitmask with sequence of set bits for final instruction
    const extendedBitmask = new Uint8Array(opcodeBitmask.length + 25)
    extendedBitmask.set(opcodeBitmask)
    extendedBitmask.fill(1, opcodeBitmask.length)

    // Find next set bit starting from i+1
    for (let j: i32 = 1; j <= 24; j++) {
      const bitIndex = instructionIndex + j
      if (
        bitIndex < extendedBitmask.length &&
        extendedBitmask[bitIndex] === 1
      ) {
        return j - 1
      }
    }

    return 24 // Maximum skip distance
  }

  /**
   * Parse a program using Gray Paper specification (with bitmask)
   *
   * This method decodes the program blob and uses the opcode bitmask
   * to determine instruction boundaries.
   *
   * @param programBlob - The encoded program blob
   * @returns Parsed instructions and errors
   */
  parseProgram(programBlob: Uint8Array): ParseResult {
    const instructions: PVMInstruction[] = []
    const errors: string[] = []

    // Decode the program blob
    const decoded = decodeBlob(programBlob)
    if (!decoded) {
      errors.push('Failed to decode program blob - invalid format')
      return new ParseResult(
        false,
        [] as PVMInstruction[],
        new Uint8Array(0),
        [] as u32[],
        errors,
        0,
      )
    }

    const code = decoded.code
    const bitmask = decoded.bitmask
    const jumpTable = decoded.jumpTable

    // Gray Paper pvm.tex equation: ζ ≡ c ⌢ [0, 0, . . . ]
    // Append 16 zeros to ensure no out-of-bounds access and trap behavior
    // This implements the infinite sequence of zeros as specified in the Gray Paper
    const extendedCode = new Uint8Array(code.length + 16)
    extendedCode.set(code)
    // Zeros are already initialized by Uint8Array constructor

    // Extend bitmask to cover the padded zeros (all 1s = valid opcode positions)
    // Gray Paper: "appends k with a sequence of set bits in order to ensure a well-defined result"
    const extendedBitmask = new Uint8Array(code.length + 16)
    extendedBitmask.set(bitmask)
    extendedBitmask.fill(1, bitmask.length) // Fill remaining positions with 1s

    let instructionIndex: i32 = 0

    // Parse instructions including the implicit trap from padded zeros
    // Continue until we've processed the original code length
    while (instructionIndex < extendedCode.length) {
      // Gray Paper Fskip-based parsing using extended bitmask
      // Check if current position has a valid opcode (bitmask check)
      if (
        instructionIndex >= extendedBitmask.length ||
        extendedBitmask[instructionIndex] === 0
      ) {
        // Not an opcode position, skip
        instructionIndex++
        continue
      }

      const opcode = extendedCode[instructionIndex]

      // Calculate Fskip(i) according to Gray Paper specification:
      // Fskip(i) = min(24, j ∈ N : (k ∥ {1,1,.})_{i+1+j} = 1)
      const fskip = this.skip(instructionIndex, extendedBitmask)
      const instructionLength = 1 + fskip

      // Extract operands from extended code (with zero padding)
      const operands = extendedCode.slice(
        instructionIndex + 1,
        instructionIndex + instructionLength,
      )

      const instruction = new PVMInstruction(
        opcode,
        operands,
        fskip,
        u32(instructionIndex),
      )

      instructions.push(instruction)

      // Advance to next instruction: ι' = ι + 1 + Fskip(ι)
      instructionIndex += instructionLength
    }

    return new ParseResult(
      errors.length === 0,
      instructions,
      extendedBitmask,
      jumpTable,
      errors,
      u32(code.length),
    )
  }
}

File: packages/pvm-rust/src/parser.rs
//! Program parser (mirrors assembly/parser.ts). Decode blob and parse instructions with bitmask.

use crate::codec::decode_blob;
use crate::pvm::PvmInstruction;

/// Result of parsing a program blob.
#[derive(Clone, Debug)]
pub struct ParseResult {
    pub success: bool,
    pub instructions: Vec<PvmInstruction>,
    /// Extended code (code + 16 zero bytes) for execution.
    pub extended_code: Vec<u8>,
    pub bitmask: Vec<u8>,
    pub jump_table: Vec<u32>,
    pub errors: Vec<String>,
    pub code_length: u32,
}

impl ParseResult {
    #[must_use]
    pub fn new(
        success: bool,
        instructions: Vec<PvmInstruction>,
        extended_code: Vec<u8>,
        bitmask: Vec<u8>,
        jump_table: Vec<u32>,
        errors: Vec<String>,
        code_length: u32,
    ) -> Self {
        Self {
            success,
            instructions,
            extended_code,
            bitmask,
            jump_table,
            errors,
            code_length,
        }
    }
}

/// PVM parser: deblob and Fskip-based instruction boundaries (Gray Paper 7.1–7.3).
pub struct PvmParser;

impl PvmParser {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }

    /// Fskip(i): distance in octets minus 1 to next instruction opcode.
    /// Gray Paper Eq 7.1: Fskip(i) = min(24, j ∈ N : (k ∥ {1,1,.})_{i+1+j} = 1).
    #[must_use]
    pub fn skip(&self, instruction_index: i32, opcode_bitmask: &[u8]) -> i32 {
        let _ext_len = opcode_bitmask.len() + 25;
        for j in 1..=24 {
            let bit_index = instruction_index + j;
            if bit_index < opcode_bitmask.len() as i32 {
                if opcode_bitmask[bit_index as usize] == 1 {
                    return j - 1;
                }
            } else {
                return j - 1;
            }
        }
        24
    }

    /// Parse program blob: decode blob then walk code using bitmask and Fskip.
    #[must_use]
    pub fn parse_program(&self, program_blob: &[u8]) -> ParseResult {
        let mut instructions = Vec::new();
        let mut errors = Vec::new();

        let Some(decoded) = decode_blob(program_blob) else {
            errors.push("Failed to decode program blob - invalid format".to_string());
            return ParseResult::new(false, instructions, vec![], vec![], vec![], errors, 0);
        };

        let code = &decoded.code;
        let bitmask = &decoded.bitmask;
        let jump_table = decoded.jump_table.clone();

        let code_len = code.len();
        let ext_len = code_len + 16;
        let mut extended_code = vec![0u8; ext_len];
        extended_code[..code_len].copy_from_slice(code);

        let mut extended_bitmask = vec![1u8; ext_len];
        extended_bitmask[..bitmask.len()].copy_from_slice(bitmask);

        let mut instruction_index: i32 = 0;

        while instruction_index < ext_len as i32 {
            if instruction_index >= extended_bitmask.len() as i32
                || extended_bitmask[instruction_index as usize] == 0
            {
                instruction_index += 1;
                continue;
            }

            let opcode = i32::from(extended_code[instruction_index as usize]);
            let fskip = self.skip(instruction_index, &extended_bitmask);
            let instruction_length = 1 + fskip;

            let op_end = (instruction_index + instruction_length as i32).min(ext_len as i32);
            let operands: Vec<u8> = extended_code
                [(instruction_index + 1) as usize..op_end as usize]
                .to_vec();

            let instruction = PvmInstruction::new(
                opcode,
                operands,
                fskip,
                instruction_index as u32,
            );
            instructions.push(instruction);

            instruction_index += instruction_length;
        }

        ParseResult::new(
            errors.is_empty(),
            instructions,
            extended_code,
            extended_bitmask,
            jump_table,
            errors,
            code_len as u32,
        )
    }
}

impl Default for PvmParser {
    fn default() -> Self {
        Self::new()
    }
}
